### PR TITLE
Add NESO package to repo

### PR DESCRIPTION
**Pull request by cmacmackin**
A Spack package has been written for NESO. This allows Spack to install it. It also allows us to easily ensure Spack will install all of the correct dependencies. Some modifications were made to the dependency packages to help facilitate this.

### CODE REVIEW

**Comment by jwscook** [comment]
@will-saunders-ukaea would you mind reviewing this please?

**Comment by will-saunders-ukaea** [comment]
Looking at the NESO package: I made some progress yesterday with setting up the build environment with different sycl implementations here: https://github.com/ExCALIBUR-NEPTUNE/NESO-Spack/blob/dev-neso-particles/packages/neso-particles/package.py. 

It's annoying that the intel-oneapi-compilers package doesn't provide sycl. Another issue is that if you set the compiler (for your spack spec) to %oneapi then the find_package(IntelDPCPP REQUIRED) will fail as the intel-oneapi-compilers package doesn't set the CMAKE_PREFIX_PATH when it's used as a compiler. I don't know if this ultimately matters - we've found out that for mkl linking having `-DCMAKE_CXX_COMPILER=dpcpp` is important (i.e. icpx instead of dpcpp will cause segfaults). For NESO-Particles I've experimentally set the find_package to QUIET for IntelDPCPP and simply set the CXX compiler and it seems to work. 

If we really need the compiler to be set to %oneapi and need the find_package(IntelDPCPP) to work then it looks like we're going to end up having (or work out how to avoid) installing intel-oneapi-compilers%oneapi which is messy if you end up installing intel-oneapi-compilers twice.

The Hipsycl builds seem more straightforward - we need a clean way to set the hipsycl-target and go from there.

**Comment by cmacmackin** [comment]
> If we really need the compiler to be set to %oneapi and need the find_package(IntelDPCPP) to work then it looks like we're going to end up having (or work out how to avoid) installing intel-oneapi-compilers%oneapi which is messy if you end up installing intel-oneapi-compilers twice.

There is a better way to do this, as desribed in the [spack documentation](https://spack.readthedocs.io/en/latest/getting_started.html#build-your-own-compiler). You can install the oneapi package and then tell spack to register it as a compiler in its own right.

```bash
spack install intel-oneapi-compilers
spack load intel-oneapi-compilers
spack compiler find
```

**Comment by will-saunders-ukaea** [comment]
I know, that's not the issue here. The issue is getting find_package(IntelDPCPP) to work and if we need it work. Just doing %oneapi does not set the env vars such that find_package(IntelDPCPP) works (maybe you see a way around this).

**Comment by jwscook** [comment]
Your doing all the same sensible things we did, Chris. We found these bugs and work arounds for them. We've done the work arounds because we assume they're easier than changing them oneapi spack package et al.

**Comment by cmacmackin** [comment]
Trying to build with `dpcpp` unfortunately causes problems for one of the transitive dependencies (`berkeley-db`). That package uses autotools to manage the build and there is a [known bug](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-and-GNU-Autotools/m-p/1296985) when trying to use `dpcpp` with autotools. Autotools gets a bit confused and thinks some temporary files generated during the configuration stage need to be linked into the shared object library. However, as those object files get deleted after configuration, it can't do this.

I'm trying to think of potential solutions:
- provide a wrapper for dpcpp that filters out these files, if present; awkward for users
- adjust the configuration files after the configuration step; requires editing the spack package for `berkeley-db`
- anything else?

**Comment by will-saunders-ukaea** [comment]
Is this the berkeley-db dependency of openblas? If so the neso nektar++ package can (tries to) avoid the openblas dependency with +mkl to enable MKL when using the intel tool chain. 

If this doesn't work: does berkeley-db build using icc/icpc as I doubt we need to use dpcpp for berkeley-db (seriously doubt it has a sycl interface we use - or any sycl at all).

**Comment by cmacmackin** [comment]
> Is this the berkeley-db dependency of openblas?

Nope, it's a dependency of perl, which is a dependcy of openssl, which is a dependency of cmake.

I'm not sure if I can get spack to mix dependencies compiled using icpc with those compiled using dpcpp in the same build, unfortunately.

**Comment by will-saunders-ukaea** [comment]
Does a spec like:
```
spack spec neso.nektar %oneapi ^cmake%gcc
```
Solve the problem for the time being? (seems to have a promising looking spec tree on my spack 0.18.1)

**Comment by cmacmackin** [comment]
> Does a spec like:
> 
> ```
> spack spec neso.nektar %oneapi ^cmake%gcc
> ```
> 
> Solve the problem for the time being? (seems to have a promising looking spec tree on my spack 0.18.1)
Thanks, I didn't actually know Spack would let you do that! The spec tree didn't quite work for me when I did it that way (the way I've set things up anyway), but did with

```
spack spec neso%dpcpp ^berkeley-db%gcc
```
Now to see if it will actually build successfully...

**Comment by cmacmackin** [comment]
@jwscook @will-saunders-ukaea Have either of you been able to build `numpy` with the oneAPI compilers? On my computer it fails (at link-time, I think). I strongly suspect this is just because the linger is a memory-hog and my computer runs out of RAM/swap space. I've confirmed that the same issue occurs when trying to build `numpy` with `clang` (on which oneAPI is based), though _not_ with the classic Intel compilers.

**Comment by will-saunders-ukaea** [comment]
Tried it (`spack install --reuse py-numpy%oneapi`) and some intel-clang binary got to 22GB usage before I killed it to avoid an oom crash.

**Comment by jwscook** [comment]
I haven't tried this, I confess.

**Comment by cmacmackin** [comment]
I think I have a reasonably sensible approach in place now for building with OneAPI/DPC++. I created a dummy package to represent DPC++/satisfy the SYCL dependency. It sets a bunch of environment variables etc. so that building with it in the NESO package "just works". There are still a few minor things to tidy up (possibly left for another time) and you need to make sure you have some of the right environment variables at run-time (this might be fixable from within the NESO binary). Plus you still have to compile BOOST and Numpy with GCC to avoid the memory problems already mentioned. But I think things are basically working.

File: packages/nektar/package.py [hunk 1]
```diff
@@ -83,7 +80,8 @@ def hasfeature(feature):
         args.append("-DNEKTAR_USE_SCOTCH=%s" % hasfeature("+scotch"))
         args.append("-DNEKTAR_USE_PETSC=OFF")
         args.append("-DNEKTAR_ERROR_ON_WARNINGS=OFF")
-        args.append("-DNEKTAR_USE_MKL=%s" % hasfeature("+mkl"))
+        args.append("-DNEKTAR_USE_MKL=%s" % hasfeature("%intel-oneapi-mkl"))
```
Comment (will-saunders-ukaea) [inline packages/nektar/package.py:83]: @cmacmackin Should this be `^intel-oneapi-mkl` instead?

**Comment by jwscook** [comment]
Could you augment the readme, please, do add a short copy-pasteable guide to building nektar with common variants e.g. oneapi & mkl vs hipsycl & openblas?

File: packages/nektar/package.py (hunk 1 above)
Comment (cmacmackin) [reply packages/nektar/package.py:83]: Yes, you are right.

**Comment by cmacmackin** [comment]
> Could you augment the readme, please, do add a short copy-pasteable guide to building nektar with common variants e.g. oneapi & mkl vs hipsycl & openblas?

Do you mean the README of this repo or the main NESO repo? Also, Nektar doesn't use SYCL so the oneapi vs. hipsycl consideration doesn't apply.

File: packages/nektar/package.py (hunk 1 above)
Comment (cmacmackin) [reply packages/nektar/package.py]: That is now fixed.

**Comment by jwscook** [comment]
> Do you mean the README of this repo or the main NESO repo? Also, Nektar doesn't use SYCL so the oneapi vs. hipsycl consideration doesn't apply.

I did mean this repo, but you're right, the NESO README should be brought up-to-date.

Ignore my comment about nektar and SYCL - that's just a hangover from getting it all to work before all this...

**Comment by cmacmackin** [comment]
>I did mean this repo, but you're right, the NESO README should be brought up-to-date.

I have made extensive additions to the NESO README as part of ExCALIBUR-NEPTUNE/NESO#93. If there is anything you think worth adding, probably best to add a note there.

I've expanded the README of this repo a bit now too.

**Comment by jwscook** [comment]
I'm just trying the build described by https://github.com/ExCALIBUR-NEPTUNE/NESO/pull/93 and if that's successful I reckon this is good to merge.

**Comment by jwscook** [comment]
I've successfully spack installed, built and run the tests with hipsycl.

File: README.md [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+spack install neso.nektar%oneapi ^intel-oneapi-mkl
+```
+These installations can coexist side-by-side.
+
+Note that the oneAPI compilers (and Clang, on which they are based)
+struggle to compile some dependencies like NumPy and Boost. You may
+need to specify those to be built using another compiler, such as the
+classic Intel compilers:
+```
+spack isntall neso.nektar%oneapi ^intel-oneapi-mkl ^py-numpy%intel ^boost%intel
```
Comment (will-saunders-ukaea) [inline README.md:29]: `isntall`-> `install`

Comment (cmacmackin) [reply README.md]: Amazing how often I make that typo.

**Comment by jwscook** [comment]
Building the oneapi spec with  `^intel-oneapi-mpi` and doing `export I_MPI_FABRICS=shm` to get it to build (so the compiler works at build time) and `export SYCL_DEVICE_FILTER=host` before running the tests on my gpu-less intel laptop, it all works. Phew!

**Comment by will-saunders-ukaea** [comment]
I successfully built the intel path with this: https://github.com/will-saunders-ukaea/docker-workspace/blob/11db8f41c01df143d5fcf6dad9204217634e60ee/neso-intel-test-branch/Dockerfile (I added the `^intel-oneapi-mpi` to the spec).

**Comment by cmacmackin** [comment]
As far as I'm concerned, this is ready to merge. @jwscook and/or @will-saunders-ukaea, feel free to hit the button if you are too.

**Pull request merged**